src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Helpers.cs

```diff
@@ -12,54 +12,116 @@ namespace System.Text.Json
 {
     public sealed partial class Utf8JsonWriter
     {
-        private bool HasPartialStringData => PartialStringDataLength != 0;
+        /// <summary>
+        /// Returns whether a JSON value can be written at the current position based on the current <see cref="_enclosingContainer"/>:
+        /// <list type="bullet">
+        /// <item>
+        /// <see cref="EnclosingContainerType.Array"/>: Writing a value is always allowed.
+        /// </item>
+        /// <item>
+        /// <see cref="EnclosingContainerType.Object"/>: Writing a value is allowed only if <see cref="_tokenType"/> is a property name.
+        /// Because we designed <see cref="EnclosingContainerType.Object"/> == <see cref="JsonTokenType.PropertyName"/>, we can just check for equality.
+        /// </item>
+        /// <item>
+        /// <see cref="EnclosingContainerType.None"/>: Writing a value is allowed only if <see cref="_tokenType"/> is None (only one value may be written at the root).
+        /// This case is identical to the previous case.
+        /// </item>
+        /// <item>
+        /// <see cref="EnclosingContainerType.Utf8StringSequence"/>, <see cref="EnclosingContainerType.Utf16StringSequence"/>, <see cref="EnclosingContainerType.Base64StringSequence"/>:
+        /// Writing a value is never valid and <see cref="_enclosingContainer"/> does not equal any <see cref="JsonTokenType"/> by construction.
+        /// </item>
+        /// </list>
+        /// This method performs better without short circuiting (this often gets inlined so using simple branch free code seems to have some benefits).
+        /// </summary>
+        private bool CanWriteValue => _enclosingContainer == EnclosingContainerType.Array | (byte)_enclosingContainer == (byte)_tokenType;
 
-        private void ClearPartialStringData() => PartialStringDataLength = 0;
+        private bool HasPartialStringData => _partialStringDataLength != 0;
 
-        private void ValidateEncodingDidNotChange(SegmentEncoding currentSegmentEncoding)
+        private void ClearPartialStringData() => _partialStringDataLength = 0;
+
+        private void ValidateWritingValue()
         {
-            if (PreviousSegmentEncoding != currentSegmentEncoding)
+            if (!CanWriteValue)
             {
-                ThrowHelper.ThrowInvalidOperationException_CannotMixEncodings(PreviousSegmentEncoding, currentSegmentEncoding);
+                OnValidateWritingValueFailed();
             }
         }
 
-        private void ValidateNotWithinUnfinalizedString()
+        [DoesNotReturn]
+        [MethodImpl(MethodImplOptions.NoInlining)]
```

Comment:
`NoInlining` should not be applied to throw helpers, otherwise RyuJIt will not detect the callsite as cold code.

These helpers should probably follow the pattern like `throw GetInvalidOperationException()` to make them small, but with no return, so the callsites can be optimized properly.